pkg/operation/botanist/secrets.go

```diff
@@ -243,6 +250,48 @@ func (b *Botanist) rotateKubeconfigSecrets(ctx context.Context, gardenerResource
 	return b.K8sGardenClient.Client().Patch(ctx, b.Shoot.Info, client.MergeFrom(oldObj))
 }
 
+func (b *Botanist) rotateSSHKeypairSecrets(ctx context.Context, gardenerResourceDataList *gardencorev1alpha1helper.GardenerResourceDataList) error {
+	var err error
+
+	currentSecret, ok := b.Secrets[v1beta1constants.SecretNameSSHKeyPair]
```

Comment:
This will not work, since `b.Secrets` is populated in the `DeploySecrets` function.

I think you should use the `gardenerResourceDataList` to get the current secret and handle the whole rotation. Otherwise some changes will not get saved in the `ShootState`. 

You can try something like this (notice that I delete both the old and new ssh keypair secrets from the controlplane, this is so that the old secret gets redeployed with data from the shootstate):
```go
currentSecret := gardenerResourceDataList.Get(v1beta1constants.SecretNameSSHKeyPair)
if currentSecret == nil {
	return fmt.Errorf("no Secret named %s loaded", v1beta1constants.SecretNameSSHKeyPair)
}

oldSecret := currentSecret.DeepCopy()
oldSecret.Name = v1beta1constants.SecretNameOldSSHKeyPair
gardenerResourceDataList.Upsert(oldSecret)

secrets := []string{
	v1beta1constants.SecretNameSSHKeyPair,
	v1beta1constants.SecretNameOldSSHKeyPair,
}

for _, secretName := range secrets {
	if err := b.K8sSeedClient.Client().Delete(ctx, &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: secretName, Namespace: b.Shoot.SeedNamespace}}); client.IgnoreNotFound(err) != nil {
		return err
	}
}
gardenerResourceDataList.Delete(v1beta1constants.SecretNameSSHKeyPair)

// remove operation annotation
```